stdlib/public/Cxx/CxxConvertibleToCollection.swift

```diff
@@ -0,0 +1,70 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// A C++ type that can be converted to a Swift collection.
+public protocol CxxConvertibleToCollection {
+  associatedtype RawIterator: UnsafeCxxInputIterator
+
+  /// Do not implement this function manually in Swift.
+  mutating func __beginUnsafe() -> RawIterator
+
+  /// Do not implement this function manually in Swift.
+  mutating func __endUnsafe() -> RawIterator
+}
+
+@inlinable @inline(__always)
+internal func forEachElement<C: CxxConvertibleToCollection>(
+  of c: C,
+  body: (C.RawIterator.Pointee) -> Void
+) {
+  var mutableC = c
+  withExtendedLifetime(mutableC) {
+    var rawIterator = mutableC.__beginUnsafe()
+    let endIterator = mutableC.__endUnsafe()
+    while rawIterator != endIterator {
+      body(rawIterator.pointee)
+      rawIterator = rawIterator.successor()
+    }
+  }
+}
+
+extension Array {
+  /// Creates an array containing the elements of a C++ collection.
+  ///
+  /// This initializer copies each element of the C++ collection to a new Swift
+  /// array.
+  ///
+  /// - Complexity: O(*n*), where *n* is the number of elements in the C++
+  ///   collection.
+  public init<C: CxxConvertibleToCollection>(_ c: C)
+    where C.RawIterator.Pointee == Element {
+
+    self.init()
```

Comment:
Can we reserve N elements here so we don't have to potentially allocate each time? 

(Can also be follow up. And it would be cool if we had some benchmarks to measure if there's actually an improvement from that change/others.)

Reply:
> Can we reserve N elements here so we don't have to potentially allocate each time?

We could do that for C++ RAC, however, C++ RAC are auto-conformed to `Swift.RandomAccessCollection` which brings an equivalent initializer. For non-random-access collections, we don't really know what N is until we actually iterate over the collection. For some collections iterating twice might produce different results so we can't do that I think.

> it would be cool if we had some benchmarks to measure if there's actually an improvement from that change/others

This won't be an improvement unfortunately, we're just making the copy explicit. It's a good idea to add a benchmark though 👍